File: clojure/client/tests/client_test.clj
(ns client-test (:require [clojure.test :refer [deftest is]] [convex.client :as convex]))
(deftest live-uri-and-bounded-queue
  (let [subscription (convex/->Subscription 1 "demo:state" {} (java.util.concurrent.ArrayBlockingQueue. 16) (atom true))]
    (doseq [value (range 20)] (convex/offer-newest! (:queue subscription) {:value value}))
    (is (= 4 (:value (convex/next-update subscription 1))))))
(deftest closed-client-is-rejected
  (let [client (convex/client "https://example.convex.cloud")]
    (.close client)
    (is (= :protocol (:kind (ex-data (try (convex/query client "demo:state" {}) (catch Exception error error))))))))

File: cpp/client/tests/client_test.cpp
#include "convex.hpp"
#include <cassert>
#include <iostream>
int main() {
  bool bad_url = false; try { convex::Client client("ftp://example.test"); } catch (const convex::Error&) { bad_url = true; } assert(bad_url);
  convex::Client client("https://example.test"); bool bad_args = false; try { client.query("demo:state", 1); } catch (const convex::Error&) { bad_args = true; } assert(bad_args);
  client.close(); bool closed = false; try { client.query("demo:state", {}); } catch (const convex::Error&) { closed = true; } assert(closed); std::cout << "client tests passed\n";
}
